KAFKA-15370: Support Participation in 2PC (KIP-939) (1/N)

artemlivshits opened the pull request:
This is just a mechanical change to make prepareTransitionTo method use named parameters instead of positional parameters.

artemlivshits committed:
KAFKA-15370: Support Participation in 2PC (KIP-939) (1/N)

This is just a mechanical change to make prepareTransitionTo method use
named parameters instead of positional parameters.

jolshan commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -380,50 +417,39 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
-                                  updateTimestamp: Long,
-                                  clientTransactionVersion: TransactionVersion): TxnTransitMetadata = {
+  private def prepareTransitionTo(state: TransactionState,
+                                  producerId: Long = this.producerId,
```
There was a previous effort to remove default arguments in order to not accidentally set something incorrect. I think the defaults make sense (keeping the same) so I think it makes sense, but wanted to mention this change.

artemlivshits replied:
The defaults just mean that there is no change, so if the caller doesn't use an argument, then it remains unchanged for the transition.  I think it makes code easier to read than to require every parameter to be specified: the changed stuff is easy to see from just glancing at the function call, no need to mentally parse which arguments are updated and which stay the same.

jolshan replied:
I agree with these points. Was just sharing some historical context on this (and other methods). I think it is valid to use here.

jolshan commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -380,50 +417,39 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
+                                  topicPartitions: immutable.Set[TopicPartition] = this.topicPartitions.toSet,
+                                  txnStartTimestamp: Long = this.txnStartTimestamp,
+                                  txnLastUpdateTimestamp: Long = this.txnLastUpdateTimestamp,
+                                  clientTransactionVersion: TransactionVersion = this.clientTransactionVersion): TxnTransitMetadata = {
```
Is the idea to keep clientTransactionVersion the same until each PrepareCommit/PrepareAbortCall? Or will KIP-939 introduce changes to set this in another transactional state.

artemlivshits replied:
Good question.  I don't think KIP-939 would rely on persisted value of client version, we should be able to just look at the values of the fields (nextProducerId, nextProducerEpoch) and figure out what logic needs to be done.

But it's a good point that this change actually makes the clientVersion "sticky" in the sense that once it's set it'll keep the value during the transitions.  Maybe to keep this change completely mechanical, transition to ongoing or empty should reset the version to TV_0.

jolshan replied:
Yeah, Things may also change with the change to add partitions implicitly (as we will have the client version there)

artemlivshits replied:
Adding a summary of offline conversations for visibility:
- we'll keep the existing client version in the state unless it needs to be changed
- we'll update the `completeTransictionTo` to always copy all fields to eliminate a second decision point for which fields are toing to be update; having a second decision point is confusing and can lead to bugs

rreddy-22 commented:
Thanks for the PR, makes things much more clear and readable!

artemlivshits committed:
Merge branch 'trunk' into KIP-939-refactor

artemlivshits committed:
KAFKA-15370: Support Participation in 2PC (KIP-939) (1/N)

Update completeTransitionTo to fully take the state from
transitMetadata, this eliminates a redundant decision point.

artemlivshits committed:
Merge branch 'trunk' into KIP-939-refactor

artemlivshits committed:
KAFKA-15370: Support Participation in 2PC (KIP-939) (1/N)

Fix unit tests to reflect changed logic (always copy transition data)

artemlivshits commented on core/src/test/scala/unit/kafka/coordinator/transaction/TransactionMetadataTest.scala:
```diff
@@ -204,7 +204,7 @@ class TransactionMetadataTest {
     assertEquals(producerEpoch + 1, txnMetadata.producerEpoch)
     assertEquals(producerEpoch, txnMetadata.lastProducerEpoch)
-    assertEquals(1L, txnMetadata.txnStartTimestamp)
+    assertEquals(-1L, txnMetadata.txnStartTimestamp)
```
This logic is changed (now we unconditionally copy data from the transitMetadata when we complete transition, so original value of 1 is not preserved), so the unit test is changed to codify the new logic.  Looking at the history, it doesn't seem like there was a significance in the logic, it's just the UT codified some behavior, and now it codifies a different behavior.

artemlivshits commented on core/src/test/scala/unit/kafka/coordinator/transaction/TransactionMetadataTest.scala:
```diff
@@ -231,7 +231,7 @@ class TransactionMetadataTest {
     assertEquals(producerEpoch, txnMetadata.lastProducerEpoch)
     assertEquals(0, txnMetadata.producerEpoch)
-    assertEquals(1L, txnMetadata.txnStartTimestamp)
+    assertEquals(-1L, txnMetadata.txnStartTimestamp)
```
Unconditionally copy data from the transitMetadata when we complete transition.

artemlivshits committed:
Merge branch 'trunk' into KIP-939-refactor

jolshan commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -329,8 +347,12 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
+      state = Ongoing,
+      topicPartitions = (topicPartitions ++ addedTopicPartitions),
+      txnStartTimestamp = newTxnStartTimestamp,
+      txnLastUpdateTimestamp = updateTimestamp
```
should we include clientTransactionVersion here since it is now passed to this method too?

artemlivshits committed:
KAFKA-15370: Support Participation in 2PC (KIP-939) (1/N)

Address review feedback.

jolshan reviewed (approved):
Thanks!
